From 411b0ab494847eb72e0a5925cafa59ffe6818d62 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:19:35 -0400
Subject: [fs] sunrpc: Handle EPIPE in xprt_connect_status

Message-id: <1411050057-12417-13-git-send-email-steved@redhat.com>
Patchwork-id: 96190
O-Subject: [RHEL7.1 PATCH 12/94] SUNRPC: Handle EPIPE in xprt_connect_status
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

The callback handler xs_error_report() can end up propagating an EPIPE
error by means of the call to xprt_wake_pending_tasks(). Ensure that
xprt_connect_status() does not automatically convert this into an
EIO error.

Reported-by: Weston Andros Adamson <dros@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 2fc193cf924ea6eb74f6a0cf73b94b2e62938ae5
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/clnt.c b/net/sunrpc/clnt.c
index 575e63f..488ddee 100644
--- a/net/sunrpc/clnt.c
+++ b/net/sunrpc/clnt.c
@@ -1814,6 +1814,7 @@ call_connect_status(struct rpc_task *task)
  case -ENETUNREACH:
  case -EHOSTUNREACH:
  case -ENOBUFS:
+ case -EPIPE:
   if (RPC_IS_SOFTCONN(task))
    break;
   /* retry with existing socket, after a delay */
diff --git a/net/sunrpc/xprt.c b/net/sunrpc/xprt.c
index 6563acf..5be25bb 100644
--- a/net/sunrpc/xprt.c
+++ b/net/sunrpc/xprt.c
@@ -744,6 +744,7 @@ static void xprt_connect_status(struct rpc_task *task)
  case -ECONNABORTED:
  case -ENETUNREACH:
  case -EHOSTUNREACH:
+ case -EPIPE:
  case -EAGAIN:
   dprintk("RPC: %5u xprt_connect_status: retrying\n", task->tk_pid);
   break;
-- 
1.7.1